[HUDI-1089] Refactor hudi-client to support multi-engine

wangxianghu opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request

*Refactor hudi-client to support multi-engine*

## Brief change log

  - *Refactor hudi-client to support multi-engine*

## Verify this pull request

*This pull request is already covered by existing tests*.

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

Mathieu1124 commented:
Hi, @vinothchandar @yanghua @leesf  as the refactor is finished, I have filed a Jira ticket to track this work,
please review this on this pr :)

leesf commented:
> Hi, @vinothchandar @yanghua @leesf as the refactor is finished, I have filed a Jira ticket to track this work,
> please review the refactor work on this pr :)

ack. @Mathieu1124 pls check travis failure.

vinothchandar commented:
@leesf @Mathieu1124 @lw309637554 so this replaces #1727 right?

vinothchandar commented:
Good to get @n3nash 's review here as well to make sure we are not breaking anything for the RDD client users..

Mathieu1124 commented:
> @leesf @Mathieu1124 @lw309637554 so this replaces #1727 right?

yes, https://github.com/apache/hudi/pull/1727  can be closed now

Mathieu1124 commented:
> > Hi, @vinothchandar @yanghua @leesf as the refactor is finished, I have filed a Jira ticket to track this work,
> > please review the refactor work on this pr :)
> 
> ack. @Mathieu1124 pls check travis failure.

copy, have resolved the ci failure and conflicts with master, will push it after work.

Mathieu1124 commented:
Hi, @vinothchandar @yanghua @leesf @n3nash, ci is green, this pr is ready for review now :)

yanghua commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/storage/HoodieParquetWriter.java:
```diff
@@ -40,7 +39,7 @@
  * the current file can take more records with the <code>canWrite()</code>
  */
-public class HoodieParquetWriter<T extends HoodieRecordPayload, R extends IndexedRecord>
+public class HoodieParquetWriter<R extends IndexedRecord>
```
Why do we need to change this class?

Mathieu1124 replied:
> Why do we need to change this class?

The Generic "T" is useless in this class, and it causes some generic problems in the abstraction, so I removed it.

vinothchandar commented:
@Mathieu1124 @leesf can you please share any tests you may have done in your own environment to ensure existing functionality is in tact.. This is a major signal we may not completely get with a PR review

Mathieu1124 commented on hudi-cli/src/main/java/org/apache/hudi/cli/commands/CompactionCommand.java:
```diff
@@ -593,8 +592,8 @@ public String repairCompaction(
-  private String getRenamesToBePrinted(List<RenameOpResult> res, Integer limit, String sortByField, boolean descending,
-      boolean headerOnly, String operation) {
+  private String getRenamesToBePrinted(List<BaseCompactionAdminClient.RenameOpResult> res, Integer limit, String sortByField, boolean descending,
+                                       boolean headerOnly, String operation) {
```
Hi, @yanghua thanks for your review. I am not sure which one is right either, I will roll back these style issues just to keep as same as before.

Mathieu1124 commented:
> @Mathieu1124 @leesf can you please share any tests you may have done in your own environment to ensure existing functionality is in tact.. This is a major signal we may not completely get with a PR review

My test is limited, just all the unit tests in source code, and all the demos in the Quick-Start Guide. I am planning to test it in docker env.

henrywu2019 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/common/HoodieEngineContext.java:
```diff
@@ -0,0 +1,48 @@
+  /**
+   * A wrapped hadoop configuration which can be serialized.
+   */
+  private SerializableConfiguration hadoopConf;
```
Just bump into this... Since this is a generic engine context, will it be better to use a generic name like `engineConfig`?

Mathieu1124 replied:
> 
> 
> Just bump into this... Since this is a generic engine context, will it be better to use a generic name like `engineConfig`?

Hi, henry thanks for your review. This class holds more than config stuff(your can see its child class `HoodieSparkEngineContext`),  maybe context is better, WDYT？

Mathieu1124 replied:
![图片](https://user-images.githubusercontent.com/49835526/87910984-76cc5400-ca9d-11ea-8279-b9a37b249e7b.png)

henrywu2019 replied:
Oh...What I meant is at line 32 the name `hadoopConf`, not the class name, which implies `hadoop`. I bumped into this searching for Flink support from HUDI and this PR looks a big step moving in that direction. Thanks tons @Mathieu1124 and definitely @vinothchandar as well.

vinothchandar commented:
@satishkotha @nbalajee @prashantwason @modi95  please take a look as well.

vinothchandar commented:
@Mathieu1124 , @leesf  : @n3nash said he is half way through reviewing. I took another pass and this seems low risk enough for us to merge for 0.6.0. 

We have some large PRs pending though #1702  #1678  #1834  . I would like to merge those and then rework this a bit on top of this. How painful do you think the rebase would be? (I can help as much as I can as well). Does this sound like a good plan to you

Mathieu1124 commented:
> @Mathieu1124 , @leesf : @n3nash said he is half way through reviewing. I took another pass and this seems low risk enough for us to merge for 0.6.0.
> 
> We have some large PRs pending though #1702 #1678 #1834 . I would like to merge those and then rework this a bit on top of this. How painful do you think the rebase would be? (I can help as much as I can as well). Does this sound like a good plan to you
@vinothchandar, I have taken a quick pass about these three PRs above, can't say that'll be little work, but I am ok with this plan because that three PRs are based on the same base, and leaving this PR at the last could greatly reduce their workload on rebasing and gives more time for us to test this PR.

vinothchandar commented:
@Mathieu1124 thanks for being awesome and understanding!. I will help out as much as I can as well. In fact, smoothly landing these large PRs is my sole focus now.

wangxianghu commented:
Hi, @vinothchandar @yanghua @leesf @n3nash Sorry for the delay, This PR is ready for review now :)
I have tested this PR  by executing the insert, update, query, delete demo in my local, It works ok. 
I know this test is far from enough, this refactor is so huge, It surely has some problems that I didn't find now.
So, in order to finish it ASAP, would you mind start to review it now, and I will focus on testing it with more scenes, and improving it.

Thanks

vinothchandar commented:
@wangxianghu sounds good. Thanks for this monumental effort! 

Will start in the earliest. can we get the CI to pass ?

wangxianghu commented:
> @wangxianghu sounds good. Thanks for this monumental effort!
> 
> Will start in the earliest. can we get the CI to pass ?

Ack, will check it tonight

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/hbase/BaseHoodieHBaseIndex.java:
```diff
@@ -0,0 +1,295 @@
+/**
+ * Hoodie Index implementation backed by HBase.
+ */
+public abstract class BaseHoodieHBaseIndex<T extends HoodieRecordPayload, I, K, O, P> extends HoodieIndex<T, I, K, O, P> {
```
are there any code changes here, i.e logic changes?

wangxianghu replied:
> are there any code changes here, i.e logic changes?

nothing changed

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/KeyGeneratorInterface.java:
```diff
@@ -34,8 +33,4 @@
 
   List<String> getRecordKeyFieldNames();
 
-  String getRecordKey(Row row);
```
we should make sure there are no backwards incompatible changes to the key generator interface

wangxianghu replied:
> we should make sure there are no backwards incompatible changes to the key generator interface

Yes, I moved it to `SparkKeyGeneratorInterface`

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseMergeHelper.java:
```diff
@@ -161,11 +108,11 @@ private static GenericRecord transformRecordBasedOnNewSchema(GenericDatumReader<
+  static class UpdateHandler extends BoundedInMemoryQueueConsumer<GenericRecord, Void> {
 
-    private final HoodieMergeHandle upsertHandle;
+    private final HoodieWriteHandle upsertHandle;
```
why is this no longer a mergeHandle?

wangxianghu replied:
> why is this no longer a mergeHandle?

`HoodieWriteHandle` is spark-free, while `HoodieMergeHandle` is not. To abstract `MergeHelper`, the variables it holds should be spark-free too

wangxianghu replied:
> why is this no longer a mergeHandle?

With `parallelDo` method introduced in, this change is no longer needed.  rollback already

vinothchandar reviewed (commented):
I assumed you had not made any logic related changes in this PR. Can your confirm/call out places where you had to change logic. 

Kind of midway there in review. Changes seem like purely refactoring and templatizing. Simpler the better in this PR. We can optimize/improve each API (base/child class contracts) over time. 
I will look at the package structure more closely next.

wangxianghu commented:
Hi, @vinothchandar thanks for your review. I'll address your concerns ASAP.
The CI failure seems caused by timeout(50min),  can you help figure this out, I don't know how to adjust this threshold.

vinothchandar commented on style/checkstyle.xml:
```diff
@@ -62,7 +62,7 @@
         </module>
         <module name="LineLength">
-            <property name="max" value="200"/>
+            <property name="max" value="500"/>
```
let's discuss this in a separate PR? 500 is a really large threshold

wangxianghu replied:
> let's discuss this in a separate PR? 500 is a really large threshold

very sorry to say this ...
According to the current abstraction，some declaration of the method is longer than 200.
such as `org.apache.hudi.table.action.compact.HoodieSparkMergeOnReadTableCompactor#generateCompactionPlan`
It is 357 characters long

vinothchandar commented on hudi-client/pom.xml:
```diff
@@ -68,6 +107,12 @@
   <dependencies>
+    <!-- Scala -->
+    <dependency>
+      <groupId>org.scala-lang</groupId>
```
should we limit scala to just the spark module?

wangxianghu replied:
> should we limit scala to just the spark module?
Yes, it is better. can we do it in another PR?
because although some classes have nothing to do with spark, it used `scala.Tuple2`, so scala is still needed. 
we can replace it with 'org.apache.hudi.common.util.collection.Pair'
WDYT?

vinothchandar replied:
fair. let me take a closer look

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/testutils/HoodieClientTestUtils.java:
```diff
@@ -81,7 +82,9 @@
     SparkConf sparkConf = new SparkConf().setAppName(appName)
-        .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer").setMaster("local[8]");
+        .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+        .set("spark.driver.host","localhost")
```
why was this change required?

wangxianghu replied:
> why was this change required?

I have rolled back this. 
The unit test is not runnable in my local yesterday, but ok now... weird

vinothchandar reviewed (commented):
One more pass. 

@wangxianghu do the tests pass locally? 50 min is the travis limit, if its consistently exceeding that limit, we need to understand why and fix it. I can help unblock once you help me with the details. 

@wangxianghu can you confirm that there are no logic changes in this PR? High level it seems ok to me. Hardest/time consuming part of the review is actually, comparing line by line for any changes in the base/child classes everywhere. If you can give me some guidance, I can finish the review quickly

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -716,32 +674,97 @@ private void rollbackPendingCommits() {
+  public abstract AsyncCleanerService startAsyncCleaningIfEnabled(AbstractHoodieWriteClient<T, I, K, O, P> client, String instantTime);
+
+  @Override
+  public void close() {
```
need to ensure the ordering of closing resources is the same as before/

wangxianghu replied:
> need to ensure the ordering of closing resources is the same as before/

Yes, they are the same.

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/HoodieSparkWriteClient.java:
```diff
@@ -0,0 +1,360 @@
+import java.util.List;
+import java.util.Map;
+
+public class HoodieSparkWriteClient<T extends HoodieRecordPayload> extends AbstractHoodieWriteClient<T,
```
Let's name this `SparkRDDWriteClient` ?

wangxianghu replied:
> Let's name this `SparkRDDWriteClient` ?

done

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/HoodieSparkWriteClient.java:
```diff
@@ -0,0 +1,360 @@
+    HoodieSparkTable table = (HoodieSparkTable) getTableAndInitCtx(WriteOperationType.UPSERT, instantTime);
+    table.validateUpsertSchema();
+    setOperationType(WriteOperationType.UPSERT);
+    startAsyncCleaningIfEnabled(this, instantTime);
```
why are we not hanging onto the returned object?

wangxianghu replied:
> why are we not hanging onto the returned object?

my bad. done

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/HoodieSparkWriteClient.java:
```diff
@@ -0,0 +1,360 @@
+    HoodieSparkTable table = (HoodieSparkTable) getTableAndInitCtx(WriteOperationType.UPSERT_PREPPED, instantTime);
+    table.validateUpsertSchema();
+    setOperationType(WriteOperationType.UPSERT_PREPPED);
+    startAsyncCleaningIfEnabled(this, instantTime);
```
same here and everywhere else.

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/embedded/SparkEmbeddedTimelineService.java:
```diff
@@ -0,0 +1,51 @@
+  @Override
+  public void setHostAddrFromContext(HoodieEngineContext context) {
+    SparkConf sparkConf = HoodieSparkEngineContext.getSparkContext(context).getConf();
+    String hostAddr = sparkConf.get("spark.driver.host", null);
```
I think we can eliminate the need for breaking this up into spark vs non-spark, by just passing in the host. This class does not make much sense being broken up.

wangxianghu replied:
> I think we can eliminate the need for breaking this up into spark vs non-spark, by just passing in the host. This class does not make much sense being broken up.

done, add `hoodie.embed.timeline.server.host` to `HoodieWriteConfig`,  it can be obtained via method `getEmbeddedServerHost()`

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/hbase/HoodieSparkHBaseIndex.java:
```diff
@@ -18,169 +18,60 @@
-    this.hBaseIndexQPSResourceAllocator = createQPSResourceAllocator(this.config);
-  }
-
-  public HBaseIndexQPSResourceAllocator createQPSResourceAllocator(HoodieWriteConfig config) {
```
note to self: make sure these methods are now in the base class

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/simple/HoodieSparkGlobalSimpleIndex.java:
```diff
@@ -71,43 +75,14 @@ public HoodieGlobalSimpleIndex(HoodieWriteConfig config) {
-  /**
-   * Load all files for all partitions as <Partition, filename> pair RDD.
-   */
-  protected List<Pair<String, HoodieBaseFile>> getAllBaseFilesInTable(final JavaSparkContext jsc, final HoodieTable hoodieTable) {
```
note to self: make sure these methods are in the base class now

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkMergeHandle.java:
```diff
@@ -54,9 +60,9 @@
 
 @SuppressWarnings("Duplicates")
-public class HoodieMergeHandle<T extends HoodieRecordPayload> extends HoodieWriteHandle<T> {
+public class HoodieSparkMergeHandle<T extends HoodieRecordPayload> extends HoodieWriteHandle<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, JavaPairRDD<HoodieKey, Option<Pair<String, String>>>> {
```
at the MergeHandle level, we need not introduce any notion of RDDs. the `io` package should be free of spark already. All we need to do is to pass in the taskContextSupplier correctly? This is a large outstanding issue we need to resolve

wangxianghu replied:
> at the MergeHandle level, we need not introduce any notion of RDDs. the `io` package should be free of spark already. All we need to do is to pass in the taskContextSupplier correctly? This is a large outstanding issue we need to resolve

Actually not yet. https://github.com/apache/hudi/pull/1756 added support for rollbacks using marker files, and `MarkerFiles` is spark related.

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkMergeHandle.java:
```diff
@@ -71,34 +77,25 @@
     init(fileId, this.partitionPath, dataFileToBeMerged);
   }
 
-  @Override
```
please refrain from moving methods around within the file. it makes life hard during review :(

wangxianghu replied:
> please refrain from moving methods around within the file. it makes life hard during review :(

sorry for the inconvenient, let me see what I can do to avoid this :)

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/SparkAppendHandleFactory.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Factory to create {@link HoodieSparkAppendHandle}.
+ */
+public class SparkAppendHandleFactory<T extends HoodieRecordPayload> extends WriteHandleFactory<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, JavaPairRDD<HoodieKey, Option<Pair<String, String>>>> {
```
same here. we need to make sure these factory methods don't have spark vs non-spark versions

wangxianghu replied:
> same here. we need to make sure these factory methods don't have spark vs non-spark versions

done

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/SparkCreateHandleFactory.java:
```diff
@@ -0,0 +1,46 @@
+public class SparkCreateHandleFactory<T extends HoodieRecordPayload> extends WriteHandleFactory<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, JavaPairRDD<HoodieKey, Option<Pair<String, String>>>> {
+
+  @Override
+  public HoodieSparkCreateHandle create(final HoodieWriteConfig hoodieConfig,
```
same. is there a way to not make these spark specific

wangxianghu replied:
> same. is there a way to not make these spark specific

I'll give a try

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/SparkWorkloadProfile.java:
```diff
@@ -22,49 +22,22 @@
-    this.partitionPathStatMap = new HashMap<>();
-    this.globalStat = new WorkloadStat();
-    buildProfile();
+public class SparkWorkloadProfile<T extends HoodieRecordPayload> extends BaseWorkloadProfile<JavaRDD<HoodieRecord<T>>> {
```
we can actually try and keep this generic and just pass in what we need from `taggedRecords` to constructor instead of the entire thing

wangxianghu replied:
> we can actually try and keep this generic and just pass in what we need from `taggedRecords` to constructor instead of the entire thing

done

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/bootstrap/SparkBootstrapCommitActionExecutor.java:
```diff
@@ -77,34 +81,44 @@
   }
 
+  @Override
+  public HoodieWriteMetadata<JavaRDD<WriteStatus>> execute(JavaRDD<HoodieRecord<T>> inputRecordsRDD) {
```
hmmm? why do we return null here

wangxianghu replied:
> hmmm? why do we return null here

`BootstrapCommitActionExecutor` dose not need this method actually, inherited from its parent class.

vinothchandar reviewed (commented):
Have to review the action executor code more closely, package structure still pending.. 

But we should have enough now, to make another pass

vinothchandar commented:
@wangxianghu the issue with the tests is that, now most of the tests are moved to hudi-spark-client. previously we had split tests into hudi-client and others. We need to edit `travis.yml` to adjust the splits again

wangxianghu commented on hudi-client/hudi-client-common/pom.xml:
```diff
@@ -0,0 +1,44 @@
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>hudi-client-common</artifactId>
```
have moved dependencies from hudi-client to hudi-client-common

wangxianghu commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/asyc/AbstractAsyncService.java:
```diff
@@ -16,7 +16,7 @@
  */
 
-package org.apache.hudi.async;
+package org.apache.hudi.asyc;
```
done

wangxianghu commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndex.java:
```diff
@@ -21,94 +21,52 @@
    * Looks up the index and tags each incoming record with a location of a file that contains the row (if it is actually
    * present).
    */
   @PublicAPIMethod(maturity = ApiMaturityLevel.STABLE)
```
> these annotations needs to moved over to a `SparkHoodieIndex` class? it will be hard for end developers to program against `HoodieIndex` directly anymore. This is a general point actually. The current public APIs should all be annotated against the Spark child classes. wdyt?

good idea, done

wangxianghu commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/RollbackUtils.java:
```diff
@@ -0,0 +1,134 @@
+   * @param shouldAssumeDatePartitioning {@code true} if date partitioning should be assumed. {@code false} otherwise.
+   * @return {@link List} of {@link ListingBasedRollbackRequest}s thus collected.
+   */
+  public static List<ListingBasedRollbackRequest> generateRollbackRequestsByListingCOW(FileSystem fs, String basePath, boolean shouldAssumeDatePartitioning) {
```
> the MOR equivalent method got moved I guess

Yes,  MOR equivalent method moved to `SparkRollbackUtils`.

wangxianghu commented on hudi-spark/src/main/scala/org/apache/hudi/IncrementalRelation.scala:
```diff
@@ -64,8 +64,7 @@ class IncrementalRelation(val sqlContext: SQLContext,
   // TODO : Figure out a valid HoodieWriteConfig
-  private val hoodieTable = HoodieTable.create(metaClient, HoodieWriteConfig.newBuilder().withPath(basePath).build(),
-    sqlContext.sparkContext.hadoopConfiguration)
+  private val hoodieTable = HoodieSparkTable.create(HoodieWriteConfig.newBuilder().withPath(basePath).build(), new HoodieSparkEngineContext(new JavaSparkContext(sqlContext.sparkContext)) ,metaClient)
```
> nit: no space before `,` and space after `, metaClient`

done

wangxianghu commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bloom/BaseHoodieBloomIndex.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Indexing mechanism based on bloom filter. Each parquet file includes its row_key bloom filter in its metadata.
+ */
+public abstract class BaseHoodieBloomIndex<T extends HoodieRecordPayload, I, K, O, P> extends HoodieIndex<T, I, K, O, P> {
```
> I suggest introducing a `SparkHoodieIndex` base class

done

wangxianghu commented on hudi-spark/src/main/java/org/apache/hudi/bootstrap/SparkParquetBootstrapDataProvider.java:
```diff
@@ -43,18 +43,18 @@
 
   @Override
-  public JavaRDD<HoodieRecord> generateInputRecordRDD(String tableName, String sourceBasePath,
+  public JavaRDD<HoodieRecord> generateInputRecord(String tableName, String sourceBasePath,
```
> rename: generateInputRecords

done

wangxianghu commented:
> One more pass.
> 
> @wangxianghu do the tests pass locally? 50 min is the travis limit, if its consistently exceeding that limit, we need to understand why and fix it. I can help unblock once you help me with the details.
> 
> @wangxianghu can you confirm that there are no logic changes in this PR? High level it seems ok to me. Hardest/time consuming part of the review is actually, comparing line by line for any changes in the base/child classes everywhere. If you can give me some guidance, I can finish the review quickly
Hi @vinothchandar, there are no logic changes in this PR, just as you said purely refactoring and templatizing.
The unit test in my local is ok.
I'll give a list of the methods moved between base/child classes.

wangxianghu commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -18,120 +18,195 @@
-   * @return SparkConf
+   * Commit changes performed at the given instantTime marker.
    */
-  public static SparkConf registerClasses(SparkConf conf) {
```
> guessing this is all moved to spark client now?

yes, In `SparkRDDWriteClient` now

wangxianghu commented:
> @wangxianghu the issue with the tests is that, now most of the tests are moved to hudi-spark-client. previously we had split tests into hudi-client and others. We need to edit `travis.yml` to adjust the splits again

@vinothchandar could you please help me edit travis.yml to adjust the splits .. I am not familiar with that
thanks :)

wangxianghu commented:
@vinothchandar @yanghua @leesf  The ci passed now

yanghua commented:
@leesf Thanks for your awesome work. Can you squash these commits for the subsequent review?

vinothchandar commented:
@wangxianghu That's awesome. Will resume the Review again this weekend

vinothchandar commented:
I can help address the remaining feedback. I will push a small diff today/tmrw.
Overall, looks like a reasonable start. 

The major feedback I still have is the following  

>would a parallelDo(func, parallelism) method in HoodieEngineContext help us avoid a lot of base/child class duplication of logic like this?

Lot of usages are like `jsc.parallelize(list, parallelism).map(func)` , which all require a base-child class now. I am wondering if its easier to take those usages alone and implement as `engineContext.parallelDo(list, func, parallelism)`. This can be the lowest common denominator across Spark/Flink etc. We can avoid splitting a good chunk of classes if we do this IMO. If this is interesting, and we agree, I can try to quantify.

wangxianghu commented:
> I can help address the remaining feedback. I will push a small diff today/tmrw.
> Overall, looks like a reasonable start.
> 
> The major feedback I still have is the following
> 
> > would a parallelDo(func, parallelism) method in HoodieEngineContext help us avoid a lot of base/child class duplication of logic like this?
> 
> Lot of usages are like `jsc.parallelize(list, parallelism).map(func)` , which all require a base-child class now. I am wondering if its easier to take those usages alone and implement as `engineContext.parallelDo(list, func, parallelism)`. This can be the lowest common denominator across Spark/Flink etc. We can avoid splitting a good chunk of classes if we do this IMO. If this is interesting, and we agree, I can try to quantify.

@vinothchandar thanks for your help, I think introducing a method to replace `jsc.parallelize(list, parallelism).map(func)` is a good idea. I'll give a try about this.

wangxianghu commented:
> I can help address the remaining feedback. I will push a small diff today/tmrw.
> Overall, looks like a reasonable start.
> 
> The major feedback I still have is the following
> 
> > would a parallelDo(func, parallelism) method in HoodieEngineContext help us avoid a lot of base/child class duplication of logic like this?
> 
> Lot of usages are like `jsc.parallelize(list, parallelism).map(func)` , which all require a base-child class now. I am wondering if its easier to take those usages alone and implement as `engineContext.parallelDo(list, func, parallelism)`. This can be the lowest common denominator across Spark/Flink etc. We can avoid splitting a good chunk of classes if we do this IMO. If this is interesting, and we agree, I can try to quantify.

Hi @vinothchandar, how about this demo?
`public abstract class AbstractHoodieEngineContext<T, R, O> {

  public abstract O parallelDo(List<T> datas, Function<List<T>, R> function, int parallesiom);

}`


`public class HoodieSparkEngineContext extends AbstractHoodieEngineContext<Integer,List<String>, JavaRDD<String>> {
  private static JavaSparkContext jsc;

  static {
    SparkConf conf = new SparkConf()
        .setMaster("local[4]")
        .set("spark.driver.host","localhost")
        .setAppName("SparkFunction");
    jsc = new JavaSparkContext(conf);
  }

  @Override
  public JavaRDD<String> parallelDo(List<Integer> datas, Function<List<Integer>, List<String>> function, int parallesiom) {
    return jsc.parallelize(function.apply(datas), parallesiom);
  }
}`

`public class App {
  public static void main(String[] args) {
    // define function
    Function<List<Integer>, List<String>> function = integers -> integers.stream().map(x -> x + " ####").collect(Collectors.toList());
    // inputs
    List<Integer> inputs = Arrays.asList(1, 2, 3, 4, 5, 6, 7);
    // context
    AbstractHoodieEngineContext<Integer, List<String>, JavaRDD<String>> sparkEngineContext = new HoodieSparkEngineContext();
    // execute
    JavaRDD<String> stringJavaRdd = sparkEngineContext.parallelDo(inputs, function, 3);
    // print result
    stringJavaRdd.foreach(x -> System.out.println(x));
  }
}`

The result are as follows:
`3 ####
4 ####
1 ####
2 ####
5 ####
6 ####
7 ####`

vinothchandar commented:
That sounds promising. We cannot easily fix `T,R,O` though right at the class level. it will be different in each setting, right?  

We can use generic methods though I think and what I had in mind was something little different
if the method signature can be the following 

```
public abstract <I,O> List<O> parallelDo(List<I> data, Function<I, O> func, int parallelism); 
```

and for the Spark the implementation 

```
public <I,O> List<O> parallelDo(List<I> data, Function<I, O> func, int parallelism) {
   return jsc.parallelize(data, parallelism).map(func).collectAsList();
}
```

and the invocation 

```
engineContext.parallelDo(Arrays.asList(1,2,3), (a) -> {
      return 0;
    }, 3);

```

Can you a small Flink implementation as well and confirm this approach will work for Flink as well. Thats another key thing to answer, before we finalize the approach. 


Note that 

a) some of the code that uses `rdd.mapToPair()` etc before collecting, have to be rewritten using Java Streams. i.e 

`engineContext.parallelDo().map()` instead of `jsc.parallelize().map().mapToPair()` currently.  

b) Also we should limit the use of parallelDo() to only cases where there are no grouping/aggregations involved. for e.g if we are doing `jsc.parallelize(files, ..).map(record).reduceByKey(..)`, collecting without reducing will lead to OOMs. We can preserve what you are doing currently, for such cases.  

But the parallelDo should help up reduce the amount of code broken up (and thus reduce the effort for the flink engine). 


@wangxianghu I am happy to shepherd this PR through from this point as well. lmk 

cc @yanghua @leesf

yanghua commented:
+1 to turn to function API fashion and standard / reduce some API's behavior. We are verifying and discussing with @wangxianghu offline.

wangxianghu commented:
> That sounds promising. We cannot easily fix `T,R,O` though right at the class level. it will be different in each setting, right?
> 
> We can use generic methods though I think and what I had in mind was something little different
> if the method signature can be the following
> 
> ```
> public abstract <I,O> List<O> parallelDo(List<I> data, Function<I, O> func, int parallelism); 
> ```
> 
> and for the Spark the implementation
> 
> ```
> public <I,O> List<O> parallelDo(List<I> data, Function<I, O> func, int parallelism) {
>    return jsc.parallelize(data, parallelism).map(func).collectAsList();
> }
> ```
> 
> and the invocation
> 
> ```
> engineContext.parallelDo(Arrays.asList(1,2,3), (a) -> {
>       return 0;
>     }, 3);
> ```
> 
> Can you a small Flink implementation as well and confirm this approach will work for Flink as well. Thats another key thing to answer, before we finalize the approach.
> 
> Note that
> 
> a) some of the code that uses `rdd.mapToPair()` etc before collecting, have to be rewritten using Java Streams. i.e
> 
> `engineContext.parallelDo().map()` instead of `jsc.parallelize().map().mapToPair()` currently.
> 
> b) Also we should limit the use of parallelDo() to only cases where there are no grouping/aggregations involved. for e.g if we are doing `jsc.parallelize(files, ..).map(record).reduceByKey(..)`, collecting without reducing will lead to OOMs. We can preserve what you are doing currently, for such cases.
> 
> But the parallelDo should help up reduce the amount of code broken up (and thus reduce the effort for the flink engine).
> 
> @wangxianghu I am happy to shepherd this PR through from this point as well. lmk
> 
> cc @yanghua @leesf

Yes, I noticed that problem lately. we should use generic methods. I also agree with the a) and b) you mentioned.
In the `parallelDo` method flink engine can operate the list directly using Java Stream, I have verified that. but there is a problem:
The function used in spark map operator is `org.apache.spark.api.java.function.Function` while what flink can use is `java.util.function.Function` we should align it. 

maybe we can use `java.util.function.Function` only,  If this way, there is no need to distinguish spark and flink, there both use java stream to implement those kind operations. the generic method could be like this:
`public class HoodieEngineContext {
  public <I, O> List<O> parallelDo(List<I> data, Function<I, O> func) {
    return data.stream().map(func).collect(Collectors.toList());
  }
}`

WDYT?

vinothchandar commented:
>maybe we can use java.util.function.Function only, If this way, there is no need to distinguish spark and flink, 

I'd prefer this if possible. Makes life much simpler. We can add more overloads down the line with BiFunction etc.. as we go. 
wdyt?

vinothchandar commented:
@wangxianghu on the checkstyle change to bump up the line count to 500, I think we should revert to 200 as it is now. 
I checked out a few of the issues. they can be brought within limit, by folding like below. 
if not, we can turn off checkstyle selectively in that block?  


```

public class HoodieSparkMergeHandle<T extends HoodieRecordPayload> extends
    HoodieWriteHandle<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, JavaPairRDD<HoodieKey, Option<Pair<String, String>>>> {

```

wangxianghu commented:
> > maybe we can use java.util.function.Function only, If this way, there is no need to distinguish spark and flink,
> 
> I'd prefer this if possible. Makes life much simpler. We can add more overloads down the line with BiFunction etc.. as we go.
> wdyt?
It's good to me. I'll refactor the code with the BiFunction.

wangxianghu commented:
> @wangxianghu on the checkstyle change to bump up the line count to 500, I think we should revert to 200 as it is now.
> I checked out a few of the issues. they can be brought within limit, by folding like below.
> if not, we can turn off checkstyle selectively in that block?
> 
> ```
> 
> public class HoodieSparkMergeHandle<T extends HoodieRecordPayload> extends
>     HoodieWriteHandle<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, JavaPairRDD<HoodieKey, Option<Pair<String, String>>>> {
> ```

Good idea. I'll give a try.

vinothchandar commented:
@wangxianghu   are we set on the `parallelDo` approach. it would be nice if you or @leesf @yanghua confirm once that it works for Flink also and we can close that loop.

Like I said, also please let me know if you need help on the PR. Would love to get this landed soon :)

wangxianghu commented:
> @wangxianghu are we set on the `parallelDo` approach. it would be nice if you or @leesf @yanghua confirm once that it works for Flink also and we can close that loop.
> 
> Like I said, also please let me know if you need help on the PR. Would love to get this landed soon :)

@vinothchandar, I'm working on the 'parallelDo' approach, it works for flink too. 
I have added three functions to `HoodieEngineContext` , to do the `map`, `flatMap`, `foreach` separately. more function can be added when needed. 
```
  public <I, O> List<O> map(List<I> data, Function<I, O> func) {
    return data.stream().map(func).collect(Collectors.toList());
  }

  public <I, O> List<O> flatMap(List<I> data, Function<? super I, ? extends Stream<? extends O>> func) {
    return data.stream().flatMap(func).collect(Collectors.toList());
  }

  public <I> void foreach(List<I> data, Consumer<I> func) {
    data.forEach(func);
  }
```
Of course, it would be great if you could provide some help, I‘d like to land this pr ASAP too 
From now on , I can focus on refactoring this code with bi function, you can change anywhere you feel unsuitable, WDYT? :)

vinothchandar commented:
@wangxianghu may be we can first do a few classes with parallelDo and see how it looks? We can then even split up the remaining work

wangxianghu commented:
> @wangxianghu may be we can first do a few classes with parallelDo and see how it looks? We can then even split up the remaining work

Yes, it makes sense

wangxianghu commented:
@vinothchandar I have refactored `org.apache.hudi.table.SparkMarkerFiles` with `parallelDo ` function, it works ok in my local(`org.apache.hudi.table.TestMarkerFiles` passed),  and the Ci has passed in my repository by now.

do you have other concerns about the `parallelDo`? can I start to refactor this or with `parallelDo ` function now?

vinothchandar commented:
@wangxianghu let me check that out and circle back by your morning time/EOD PST. we can go from there. Thanks!

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/common/HoodieEngineContext.java:
```diff
@@ -0,0 +1,66 @@
+    return taskContextSupplier;
+  }
+
+  public <I, O> List<O> map(List<I> data, Function<I, O> func) {
```
I think we should leave this abstract and let the engines implement this?  even for Java. Its better to have a `HoodieJavaEngineContext`. From what I can see, this is not overridden in `HoodieSparkEngineContext` and thus we lose the parallel execution that we currently have with Spark with this change.

vinothchandar replied:
Also these APIs should take in a `parallelism` parameter, no?

wangxianghu replied:
> I think we should leave this abstract and let the engines implement this? even for Java. Its better to have a `HoodieJavaEngineContext`. From what I can see, this is not overridden in `HoodieSparkEngineContext` and thus we lose the parallel execution that we currently have with Spark with this change.

as we discussed before, parallelDo model need a function as input parameter, Unfortunately, different engines need different type function, its hard to align them in an abstract parallelDo method. so we agree to use ` java.util.function.Function` as the unified input function. in this way, there is no need to distinguish spark and flink, and the parallelism is not needed too.

vinothchandar replied:
@wangxianghu functionality wise, you are correct. it can be implemented just using Java. but, we do parallelization of different pieces of code e.g deletion of files in parallel using spark for a reason. It significantly speeds these up, for large tables. 

All I am saying is to implement the `HoodieSparkEngineContext#map` like below

```
 public <I, O> List<O> map(List<I> data, Function<I, O> func, int parallelism) {
    return javaSparkContext.parallelize(data, parallelism).map(func).collect();
 }
```

similarly for the other two methods. I don't see any issues with this. do you?

wangxianghu replied:
> @wangxianghu functionality wise, you are correct. it can be implemented just using Java. but, we do parallelization of different pieces of code e.g deletion of files in parallel using spark for a reason. It significantly speeds these up, for large tables.
> 
> All I am saying is to implement the `HoodieSparkEngineContext#map` like below
> 
> ```
>  public <I, O> List<O> map(List<I> data, Function<I, O> func, int parallelism) {
>     return javaSparkContext.parallelize(data, parallelism).map(func).collect();
>  }
> ```
> 
> similarly for the other two methods. I don't see any issues with this. do you?

I know what you mean.

vinothchandar replied:
Is it possible to take a `java.util.function.Function` and then within `HoodieSparkEngineContext#map` wrap that into a `org.apache.spark.api.java.function.Function` ?

wangxianghu replied:
> Is it possible to take a `java.util.function.Function` and then within `HoodieSparkEngineContext#map` wrap that into a `org.apache.spark.api.java.function.Function` ?

let me try

vinothchandar replied:
In Spark, there is a functional interface defined like this

```
package org.apache.spark.api.java.function;

import java.io.Serializable;

/**
 * Base interface for functions whose return types do not create special RDDs. PairFunction and
 * DoubleFunction are handled separately, to allow PairRDDs and DoubleRDDs to be constructed
 * when mapping RDDs of other types.
 */
@FunctionalInterface
public interface Function<T1, R> extends Serializable {
  R call(T1 v1) throws Exception;
}
```

vinothchandar replied:
when the use passes in a regular lambda, into `rdd.map()`, this is what it gets converted into

wangxianghu replied:
> when the use passes in a regular lambda, into `rdd.map()`, this is what it gets converted into

The serializable issue can be solved by introducing a seriableFuncition to replace `java.util.function.Function`

```
public interface SerializableFunction<I, O> extends Serializable {
  O call(I v1) throws Exception;
}
```
`HoodieEngineContext` can be
```
public abstract class HoodieEngineContext {
  public abstract  <I, O> List<O> map(List<I> data, SerializableFunction<I, O> func, int parallelism) ;
}
```

`HoodieSparkEngineContext` can be
```
public class HoodieSparkEngineContext extends HoodieEngineContext {
  private static JavaSparkContext jsc;

  // tmp
  static {
    SparkConf conf = new SparkConf()
        .setMaster("local[4]")
        .set("spark.driver.host","localhost")
        .setAppName("HoodieSparkEngineContext");

    jsc = new JavaSparkContext(conf);
  }
  
  @Override
  public <I, O> List<O> map(List<I> data, SerializableFunction<I, O> func, int parallelism) {
    return jsc.parallelize(data, parallelism).map(func::call).collect();
  }
}
```
this works :)

vinothchandar replied:
@wangxianghu This is awesome. Hopefully this can reduce the amount of code you need to write for Flink significantly. `TestMarkerFiles` seems to pass, so guess the serialization etc is working as expected. 

We can go ahead with doing more files in this approach and remerge the base/child classes back as much as possible. cc @leesf @yanghua  as well in case they have more things to add. 

cc @bvaradar as well as FYI

wangxianghu replied:
> @wangxianghu This is awesome. Hopefully this can reduce the amount of code you need to write for Flink significantly. `TestMarkerFiles` seems to pass, so guess the serialization etc is working as expected.
> 
> We can go ahead with doing more files in this approach and remerge the base/child classes back as much as possible. cc @leesf @yanghua as well in case they have more things to add.
> 
> cc @bvaradar as well as FYI

Yes, it also reduce tons of code in the refactoring. I'm working on it ,hope to finish it today or tomorrow

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/common/HoodieSparkEngineContext.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * A Spark engine implementation of HoodieEngineContext.
+ */
+public class HoodieSparkEngineContext extends HoodieEngineContext {
```
can we implement versiosn of `map`, `flatMap`, `forEach` here which use `javaSparkContext.parallelize()` ? It would be good to keep this PR free of any changes in terms of whether we are executing the deletes/lists in parallel or in serial.

wangxianghu replied:
> can we implement versiosn of `map`, `flatMap`, `forEach` here which use `javaSparkContext.parallelize()` ? It would be good to keep this PR free of any changes in terms of whether we are executing the deletes/lists in parallel or in serial.

done

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/SparkMarkerFiles.java:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  @Override
+  public Set<String> createdAndMergedDataPaths(HoodieEngineContext context, int parallelism) throws IOException {
```
we are not using parallelism here. This will lead to a perf regression w.r.t master.

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/SparkMarkerFiles.java:
```diff
@@ -0,0 +1,124 @@
+import static org.apache.hudi.bifunction.wrapper.BiFunctionWrapper.throwingConsumerWrapper;
+import static org.apache.hudi.bifunction.wrapper.BiFunctionWrapper.throwingFlatMapWrapper;
+
+public class SparkMarkerFiles extends BaseMarkerFiles {
```
Given this file is now free of Spark, we dont have the need of breaking these into base and child classes right.

wangxianghu replied:
> Given this file is now free of Spark, we dont have the need of breaking these into base and child classes right.

Yes, this is an example to show you the bi function, if you agree with this implementation, I'll rollback them in one class

vinothchandar reviewed (commented):
My primary motive of suggesting parallelDo model, is to avoid splitting the classes and still reap benefits of parallel execution, provided by each engine. I don't think we are realizing them, as this stage yet. Please let me know your thoughts.

wangxianghu commented:
> My primary motive of suggesting parallelDo model, is to avoid splitting the classes and still reap benefits of parallel execution, provided by each engine. I don't think we are realizing them, as this stage yet. Please let me know your thoughts.

@vinothchandar The `org.apache.hudi.table.SparkMarkerFiles` is used in many places and the refactor work is huge if rollback them in one class, so I refactored the function using bi function first(without packaging them in one class)  just to show you the functional changes, thinking if you agree with this refactor then I can rollback them without splitting classes.
It is just an example.

wangxianghu commented:
@vinothchandar @yanghua @leesf 
I have refactored this pr with `parallelDo` function and fixed the checkstyle(lineLength > 200) issue
please take a look when free

yanghua commented on hudi-cli/pom.xml:
```diff
@@ -148,7 +148,14 @@
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hudi</groupId>
+      <artifactId>hudi-spark-client</artifactId>
```
Actually, I am thinking `hudi-spark-client` and `hudi-client-spark`, which is better? Given there is a module named `hudi-client-common`. wdyt? @vinothchandar

vinothchandar replied:
cc @wangxianghu probably `hudi-client-spark` is easier on the eyes?

wangxianghu replied:
> cc @wangxianghu probably `hudi-client-spark` is easier on the eyes?

I am ok with both of them :)  let's rename it to `hudi-client-spark`

yanghua replied:
So, will we keep `hudi-spark-client`? We have had a `hudi-spark` module.

wangxianghu replied:
> So, will we keep `hudi-spark-client`? We have had a `hudi-spark` module. IMHO, this naming may not seem so clear.

@vinothchandar @yanghua I have filed a pr to rename `hudi-spark-client` to `hudi-client-spark` : https://github.com/apache/hudi/pull/2139
please take a look when free

wangxianghu commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieAppendHandle.java:
```diff
@@ -134,7 +138,7 @@ private void init(HoodieRecord record) {
       writeStatus.getStat().setFileId(fileId);
-      averageRecordSize = SizeEstimator.estimate(record);
+      averageRecordSize = sizeEstimator.sizeEstimate(record);
       try {
```
Here, I used `org.apache.hudi.common.util.SizeEstimator#sizeEstimate` to replace `org.apache.spark.util.SizeEstimator#estimate` is it ok?  @vinothchandar

vinothchandar replied:
Should be okay.

wangxianghu commented:
> @wangxianghu on the checkstyle change to bump up the line count to 500, I think we should revert to 200 as it is now.
> I checked out a few of the issues. they can be brought within limit, by folding like below.
> if not, we can turn off checkstyle selectively in that block?
> 
> ```
> 
> public class HoodieSparkMergeHandle<T extends HoodieRecordPayload> extends
>     HoodieWriteHandle<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, JavaPairRDD<HoodieKey, Option<Pair<String, String>>>> {
> ```

done

wangxianghu commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AsyncCleanerService.java:
```diff
@@ -52,19 +52,6 @@ protected AsyncCleanerService(HoodieWriteClient<?> writeClient, String cleanInst
     }), executor);
   }
 
-  public static AsyncCleanerService startAsyncCleaningIfEnabled(HoodieWriteClient writeClient,
```
> this method need not have moved?

It is back now.

vinothchandar commented:
@wangxianghu Looks like we have much less class splitting now. I want to try and reduce this further if possible. 
If its alright with you, I can take over from here, make some changes and push another commit on top of yours, to try and get this across the finish line. Want to coordinate so that we are not making parallel changes,

wangxianghu commented:
> @wangxianghu Looks like we have much less class splitting now. I want to try and reduce this further if possible.
> If its alright with you, I can take over from here, make some changes and push another commit on top of yours, to try and get this across the finish line. Want to coordinate so that we are not making parallel changes,

@vinothchandar, Yeah, Of course, you can take over from here, this will greatly facilitate the process
thanks 👍

vinothchandar commented:
Alright, taking over the wheel :) thanks @wangxianghu . This is true champion effort!

vinothchandar reviewed (commented):
222/369 files now reviewed and pushed the fixes.

wangxianghu commented:
@vinothchandar, Thanks for your patience and detailed review and fix. It's a really big help!
 Do you mean I take over the rest of the work from here? :)

wangxianghu commented on hudi-cli/src/test/java/org/apache/hudi/cli/commands/TestArchivedCommitsCommand.java:
```diff
@@ -92,8 +93,9 @@ public void init() throws IOException {
     // archive
-    HoodieTimelineArchiveLog archiveLog = new HoodieTimelineArchiveLog(cfg, hadoopConf);
-    archiveLog.archiveIfRequired(jsc);
+    HoodieSparkTable table = HoodieSparkTable.create(cfg, context, metaClient);
```
> can we replace more of the code to direclty just use `HoodieTable` instead. Need to examine cases that need an explicit HoodieSparkTable

yes, make sense

vinothchandar commented:
@wangxianghu I will continue to work on this. Planning to get this landed this week. 
May need some help testing once I am there. Will ping you here!

wangxianghu commented:
> testing

Ok, please ping me when it is ready.

vinothchandar reviewed (commented):
302/372 now. Whats pending is a full review of the table and index packages.

vinothchandar commented:
rebasing against master now, its a pretty tricky one, given #2048 has introduced a new action and made changes on top of WriteClient

vinothchandar commented:
@wangxianghu @yanghua I have rebased this against master. Please take a look at my changes.

High level, we could re-use more code, but it needs an abstraction that can wrap `RDD` or `DataSet` or `DataStream` adequately and support basic operations like `.map()`, `reduceByKey()` etc. We can do this in a second pass once we have a working Flink impl.  For now this will do. 

I am trying to get the tests to pass. if they do, we could go ahead and merge

wangxianghu commented:
> @wangxianghu @yanghua I have rebased this against master. Please take a look at my changes.
> 
> High level, we could re-use more code, but it needs an abstraction that can wrap `RDD` or `DataSet` or `D

> @wangxianghu @yanghua I have rebased this against master. Please take a look at my changes.
> 
> High level, we could re-use more code, but it needs an abstraction that can wrap `RDD` or `DataSet` or `DataStream` adequately and support basic operations like `.map()`, `reduceByKey()` etc. We can do this in a second pass once we have a working Flink impl. For now this will do.
> 
> I am trying to get the tests to pass. if they do, we could go ahead and merge

Thanks, @vinothchandar, this is really great work! 
Yes, we can do more abstractions about basic `map`, `reduceByKey` methods in `HoodieEngineContext`, or some Util classes next.

vinothchandar commented:
I actually figured out that we can remove `P` altogether. since `HoodieIndex#fetchRecordLocation` is not used much outside of internal APIs. So will push a final change for that . tests are passing now

vinothchandar commented:
@wangxianghu Please help test this out if possible. Once the tests pass again, planning to merge this, morning PST 

cc @yanghua

leesf commented:
Run quickstart demo: found the warn log: 
`20/10/01 21:11:18 WARN embedded.EmbeddedTimelineService: Unable to find driver bind address from spark config`, but not found the warn log in 0.6.0 @vinothchandar @wangxianghu

vinothchandar commented:
@leesf do you see the following exception? could not understand how you ll get the other one even. 

```
LOG.info("Starting Timeline service !!");
        Option<String> hostAddr = context.getProperty(EngineProperty.EMBEDDED_SERVER_HOST);
        if (!hostAddr.isPresent()) {
          throw new HoodieException("Unable to find host address to bind timeline server to.");
        }
        timelineServer = Option.of(new EmbeddedTimelineService(context, hostAddr.get(),
            config.getClientSpecifiedViewStorageConfig()));
```

Either way, good pointer. the behavior has changed around this a bit actually. So will try and tweak and push a fix

wangxianghu commented:
> @leesf do you see the following exception? could not understand how you ll get the other one even.
> 
> ```
> LOG.info("Starting Timeline service !!");
>         Option<String> hostAddr = context.getProperty(EngineProperty.EMBEDDED_SERVER_HOST);
>         if (!hostAddr.isPresent()) {
>           throw new HoodieException("Unable to find host address to bind timeline server to.");
>         }
>         timelineServer = Option.of(new EmbeddedTimelineService(context, hostAddr.get(),
>             config.getClientSpecifiedViewStorageConfig()));
> ```
> 
> Either way, good pointer. the behavior has changed around this a bit actually. So will try and tweak and push a fix

I got this warning too. The code here seems not the same.
```
 // Run Embedded Timeline Server
        LOG.info("Starting Timeline service !!");
        Option<String> hostAddr = context.getProperty(EngineProperty.EMBEDDED_SERVER_HOST);
        timelineServer = Option.of(new EmbeddedTimelineService(context, hostAddr.orElse(null),
            config.getClientSpecifiedViewStorageConfig()));
```

vinothchandar commented:
@wangxianghu can you please test the latest commit.  To be clear, you are saying you don't get the warning on master, but get it on this branch. right?

if this round of tests pass, and you confirm, we can land from my perspective

wangxianghu commented:
> @wangxianghu can you please test the latest commit. To be clear, you are saying you don't get the warning on master, but get it on this branch. right?
> 
> if this round of tests pass, and you confirm, we can land from my perspective

Hi @vinothchandar. 

> @wangxianghu can you please test the latest commit. To be clear, you are saying you don't get the warning on master, but get it on this branch. right?
> 
> if this round of tests pass, and you confirm, we can land from my perspective

Hi @vinothchandar The warn log is still there in HUDI-1089 branch.(master is ok, no warn log)
I think we should check `embeddedTimelineServiceHostAddr` instead of `hostAddr`.

```
  private void setHostAddr(String embeddedTimelineServiceHostAddr) {
   // here we should check embeddedTimelineServiceHostAddr instead of hostAddr, hostAddr is always null
    if (hostAddr != null) {
      LOG.info("Overriding hostIp to (" + embeddedTimelineServiceHostAddr + ") found in spark-conf. It was " + this.hostAddr);
      this.hostAddr = embeddedTimelineServiceHostAddr;
    } else {
      LOG.warn("Unable to find driver bind address from spark config");
      this.hostAddr = NetworkUtils.getHostname();
    }
  }
```

wangxianghu commented:
@vinothchandar The warn log issue is fixed

vinothchandar commented:
@wangxianghu duh ofc. I understand now. Thanks for jumping in @wangxianghu !

vinothchandar commented:
@wangxianghu Just merged! Thanks again for the herculean effort. 

May be some followups could pop up. Would you be interested in taking them up? if so, I ll mention you along the way

wangxianghu commented:
> @wangxianghu Just merged! Thanks again for the herculean effort.
> 
> May be some followups could pop up. Would you be interested in taking them up? if so, I ll mention you along the way

sure, just ping me when needed

TengHuo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseCommitActionExecutor.java:
```diff
@@ -0,0 +1,159 @@
+    return true;
+  }
+
+  protected abstract Iterator<List<WriteStatus>> handleInsert(String idPfx,
+      Iterator<HoodieRecord<T>> recordItr) throws Exception;
+
+  protected abstract Iterator<List<WriteStatus>> handleUpdate(String partitionPath, String fileId,
```
May I ask if the difference of parameter `idPfx` and `fileId` between  the method `handleInsert` and `handleUpdate` is designed on some purpose?

I can understand for update operation, we need to find out the previous version of base file, so use file id to locate it.

But in the method `BaseSparkCommitActionExecutor.handleUpsertPartition`, both the method `handleInsert` and `handleUpdate` get the parameters from `binfo.fileIdPrefix`. It's really confused.

```
  protected Iterator<List<WriteStatus>> handleUpsertPartition(String instantTime, Integer partition, Iterator recordItr,
                                                              Partitioner partitioner) {
    SparkHoodiePartitioner upsertPartitioner = (SparkHoodiePartitioner) partitioner;
    BucketInfo binfo = upsertPartitioner.getBucketInfo(partition);
    BucketType btype = binfo.bucketType;
    try {
      if (btype.equals(BucketType.INSERT)) {
        return handleInsert(binfo.fileIdPrefix, recordItr);
      } else if (btype.equals(BucketType.UPDATE)) {
        return handleUpdate(binfo.partitionPath, binfo.fileIdPrefix, recordItr);
      } else {
        throw new HoodieUpsertException("Unknown bucketType " + btype + " for partition :" + partition);
      }
    } catch (Throwable t) {
      String msg = "Error upserting bucketType " + btype + " for partition :" + partition;
      LOG.error(msg, t);
      throw new HoodieUpsertException(msg, t);
    }
  }
```
